src/query/functions/src/aggregates/aggregate_unary.rs

```diff
@@ -0,0 +1,306 @@
+// Copyright 2021 Datafuse Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::alloc::Layout;
+use std::fmt::Display;
+use std::fmt::Formatter;
+use std::marker::PhantomData;
+use std::sync::Arc;
+
+use common_arrow::arrow::bitmap::Bitmap;
+use common_exception::ErrorCode;
+use common_exception::Result;
+use common_expression::types::decimal::*;
+use common_expression::types::number::*;
+use common_expression::types::DataType;
+use common_expression::types::NumberDataType;
+use common_expression::with_number_mapped_type;
+use common_expression::AggregateFunction;
+use common_expression::AggregateFunctionRef;
+use common_expression::Column;
+use common_expression::ColumnBuilder;
+use common_expression::Scalar;
+use common_expression::StateAddr;
+use ethnum::i256;
+
+use crate::aggregates::aggregate_function_factory::AggregateFunctionDescription;
+use crate::aggregates::assert_unary_arguments;
+use crate::aggregates::QuantileStateV2;
+use crate::aggregates::SkewnessStateV2;
+use crate::with_simple_no_number_mapped_type;
+
+pub trait UnaryState<T>: Send + Sync + Clone {
+    fn merge(&mut self, rhs: &Self) -> Result<()>;
+    fn merge_result(&mut self, builder: &mut ColumnBuilder) -> Result<()>;
+
+    fn accumulate(
+        &mut self,
+        columns: &[Column],
+        validity: Option<&Bitmap>,
+        input_rows: usize,
+    ) -> Result<()>;
+    fn accumulate_row(&mut self, columns: &[Column], row: usize) -> Result<()>;
+    fn accumulate_keys(
+        &self,
+        places: &[StateAddr],
+        offset: usize,
+        columns: &[Column],
+        input_rows: usize,
+    ) -> Result<()>;
+
+    fn serialize(&self, writer: &mut Vec<u8>) -> Result<()>;
+
+    fn deserialize(&self, reader: &mut &[u8]) -> Result<Self>
+    where Self: Sized;
+}
+
+pub struct AggregateUnaryFunction<S, T>
+where S: UnaryState<T>
+{
+    display_name: String,
+    state: S,
```

Comment:
_state: PhantomData<S>

Reply:
the `state` is useful now for `accumulate_keys`

Reply:
agg function is stateless, it won't keep states as field.